### DISCUSSION TITLE
What is the difference between proc, ebpf, audit, ftrace?

### DISCUSSION THREAD

**Discussion by impredicative** [evilsocket/opensnitch#569] [Q&A] [answered]
What is the difference between the various process monitor methods? These are proc, ebpf, audit, and ftrace. Also, which is the default? Why should I choose one over the other?

**Comment by gustavo-iniguez-goya** [comment, marked as answer]
Hi @impredicative ,

- [eBPF](https://github.com/evilsocket/opensnitch/wiki/monitor-method-ebpf) gets the PID of a connection directly from the kernel: fast and reliable.
- [Audit](https://github.com/evilsocket/opensnitch/wiki/monitor-method-auditd) gets the PID of a connection also from the kernel, but from userspace: it have some caveats, just another method.
- Proc gets the PID of a connection using netlink and parsing `/proc/` directories: slow and not always reliable. Can also be fooled.
- Ftrace is deprecated in favor of eBPF.

The default method is eBPF. In all cases, if we fail to get the PID that is trying to establish an outgoing connection, we fallback to ProcFS. Sometimes eBPF is not available, or Audit fails, etc... but ProcFS is always available (when it is).
